V3.0.0 存算分离 FE leader起不来 be正常 另外的fe也正常但是服务不正常 升级到3.0.5 leader 起来报错如下

<!-- (At least include the following, feel free to add if you have more content) -->

### Steps to reproduce the behavior (Required)
1. CREATE TABLE '...'
2. INSERT INTO '....'
3. SELECT  '....'

### Expected behavior (Required)

### Real behavior (Required)

### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`

2023-08-16 23:26:55,710 ERROR (stateChangeExecutor|73) [GlobalStateMgr.transferToLeader():1148] failed to init journal after transfer to leader! will exit
com.google.gson.JsonSyntaxException: java.lang.IllegalStateException: Expected BEGIN_OBJECT but was STRING at line 1 column 91 path $.p.m2.
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:226) ~[spark-dpp-1.0.0.jar:?]
        at com.starrocks.persist.gson.GsonUtils$ProcessHookTypeAdapterFactory$1.read(GsonUtils.java:641) ~[starrocks-fe.jar:?]
        at com.google.gson.internal.bind.TypeAdapterRuntimeTypeWrapper.read(TypeAdapterRuntimeTypeWrapper.java:41) ~[spark-dpp-1.0.0.jar:?]
        at com.google.gson.internal.bind.MapTypeAdapterFactory$Adapter.read(MapTypeAdapterFactory.java:186) ~[spark-dpp-1.0.0.jar:?]
        at com.google.gson.internal.bind.MapTypeAdapterFactory$Adapter.read(MapTypeAdapterFactory.java:145) ~[spark-dpp-1.0.0.jar:?]
        at com.starrocks.persist.gson.GsonUtils$ProcessHookTypeAdapterFactory$1.read(GsonUtils.java:641) ~[starrocks-fe.jar:?]
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.read(ReflectiveTypeAdapterFactory.java:131) ~[spark-dpp-1.0.0.jar:?]
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:222) ~[spark-dpp-1.0.0.jar:?]
        at com.starrocks.persist.gson.GsonUtils$ProcessHookTypeAdapterFactory$1.read(GsonUtils.java:641) ~[starrocks-fe.jar:?]
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.read(ReflectiveTypeAdapterFactory.java:131) ~[spark-dpp-1.0.0.jar:?]
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:222) ~[spark-dpp-1.0.0.jar:?]
        at com.starrocks.persist.gson.GsonUtils$ProcessHookTypeAdapterFactory$1.read(GsonUtils.java:641) ~[starrocks-fe.jar:?]
        at com.google.gson.Gson.fromJson(Gson.java:963) ~[spark-dpp-1.0.0.jar:?]
        at com.google.gson.Gson.fromJson(Gson.java:928) ~[spark-dpp-1.0.0.jar:?]
        at com.google.gson.Gson.fromJson(Gson.java:877) ~[spark-dpp-1.0.0.jar:?]
        at com.google.gson.Gson.fromJson(Gson.java:848) ~[spark-dpp-1.0.0.jar:?]
        at com.starrocks.persist.UserPrivilegeCollectionInfo.read(UserPrivilegeCollectionInfo.java:75) ~[starrocks-fe.jar:?]
        at com.starrocks.journal.JournalEntity.readFields(JournalEntity.java:990) ~[starrocks-fe.jar:?]
        at com.starrocks.journal.bdbje.BDBJournalCursor.deserializeData(BDBJournalCursor.java:251) ~[starrocks-fe.jar:?]
        at com.starrocks.journal.bdbje.BDBJournalCursor.next(BDBJournalCursor.java:295) ~[starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr.replayJournalInner(GlobalStateMgr.java:2144) ~[starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr.replayJournal(GlobalStateMgr.java:2104) ~[starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr.transferToLeader(GlobalStateMgr.java:1143) ~[starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr.access$100(GlobalStateMgr.java:325) ~[starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr$1.transferToLeader(GlobalStateMgr.java:722) ~[starrocks-fe.jar:?]
        at com.starrocks.ha.StateChangeExecutor.runOneCycle(StateChangeExecutor.java:103) ~[starrocks-fe.jar:?]
        at com.starrocks.common.util.Daemon.run(Daemon.java:115) ~[starrocks-fe.jar:?]
Caused by: java.lang.IllegalStateException: Expected BEGIN_OBJECT but was STRING at line 1 column 91 path $.p.m2.
        at com.google.gson.stream.JsonReader.beginObject(JsonReader.java:384) ~[spark-dpp-1.0.0.jar:?]
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:215) ~[spark-dpp-1.0.0.jar:?]

--- a/fe/fe-core/src/main/java/com/starrocks/persist/UserPrivilegeCollectionInfo.java
+++ b/fe/fe-core/src/main/java/com/starrocks/persist/UserPrivilegeCollectionInfo.java
@@ -19,12 +19,17 @@ import com.google.gson.annotations.SerializedName;
 import com.starrocks.common.io.Text;
 import com.starrocks.common.io.Writable;
 import com.starrocks.persist.gson.GsonUtils;
+import com.starrocks.privilege.ObjectTypeDeprecate;
+import com.starrocks.privilege.PrivilegeEntry;
+import com.starrocks.privilege.UserPrivilegeCollection;
 import com.starrocks.privilege.UserPrivilegeCollectionV2;
 import com.starrocks.sql.ast.UserIdentity;
 
 import java.io.DataInput;
 import java.io.DataOutput;
 import java.io.IOException;
+import java.util.List;
+import java.util.Map;
 
 public class UserPrivilegeCollectionInfo implements Writable {
 
@@ -72,6 +77,32 @@ public class UserPrivilegeCollectionInfo implements Writable {
 
     public static UserPrivilegeCollectionInfo read(DataInput in) throws IOException {
         String json = Text.readString(in);
-        return GsonUtils.GSON.fromJson(json, UserPrivilegeCollectionInfo.class);
+        try {
+            return GsonUtils.GSON.fromJson(json, UserPrivilegeCollectionInfo.class);
+        } catch (Exception e) {
+            /*
+             * Because 3.0.3 introduced a bug, the upgrade of UserPrivilegeCollectionV2
+             * was not compatible and the original SerializedName "p" was used directly,
+             * resulting in inconsistent types after the upgrade.
+             * Fault tolerance code has been added here.
+             * If UserPrivilegeCollectionV2 cannot be parsed correctly,
+             * need to use UserPrivilegeCollectionInfoDeprecated to parse.
+             */
+            UserPrivilegeCollectionInfoDeprecated deprecated
+                    = GsonUtils.GSON.fromJson(json, UserPrivilegeCollectionInfoDeprecated.class);
+
+            UserPrivilegeCollection collectionDeprecate = deprecated.getPrivilegeCollection();
+            UserPrivilegeCollectionV2 collection = new UserPrivilegeCollectionV2();
+            collection.grantRoles(collectionDeprecate.getAllRoles());
+            collection.setDefaultRoleIds(collectionDeprecate.getDefaultRoleIds());
+
+            Map<ObjectTypeDeprecate, List<PrivilegeEntry>> m = collectionDeprecate.getTypeToPrivilegeEntryList();
+            for (Map.Entry<ObjectTypeDeprecate, List<PrivilegeEntry>> entry : m.entrySet()) {
+                collection.getTypeToPrivilegeEntryList().put(entry.getKey().toObjectType(), entry.getValue());
+            }
+
+            return new UserPrivilegeCollectionInfo(deprecated.getUserIdentity(), collection,
+                    deprecated.pluginId, deprecated.pluginVersion);
+        }
     }
 }

--- a/fe/fe-core/src/main/java/com/starrocks/persist/UserPrivilegeCollectionInfoDeprecated.java
+++ b/fe/fe-core/src/main/java/com/starrocks/persist/UserPrivilegeCollectionInfoDeprecated.java
@@ -1,0 +1,76 @@
+// Copyright 2021-present StarRocks, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.starrocks.persist;
+
+import com.google.gson.annotations.SerializedName;
+import com.starrocks.common.io.Text;
+import com.starrocks.common.io.Writable;
+import com.starrocks.persist.gson.GsonUtils;
+import com.starrocks.privilege.UserPrivilegeCollection;
+import com.starrocks.sql.ast.UserIdentity;
+
+import java.io.DataInput;
+import java.io.DataOutput;
+import java.io.IOException;
+
+public class UserPrivilegeCollectionInfoDeprecated implements Writable {
+
+    @SerializedName(value = "i")
+    public short pluginId;
+    @SerializedName(value = "v")
+    public short pluginVersion;
+    @SerializedName(value = "u")
+    private UserIdentity userIdentity;
+
+    @SerializedName(value = "p")
+    private UserPrivilegeCollection privilegeCollection;
+
+    public UserPrivilegeCollectionInfoDeprecated(
+            UserIdentity userIdentity,
+            UserPrivilegeCollection userPrivilegeCollection,
+            short pluginId,
+            short pluginVersion) {
+        this.userIdentity = userIdentity;
+        this.privilegeCollection = userPrivilegeCollection;
+        this.pluginId = pluginId;
+        this.pluginVersion = pluginVersion;
+    }
+
+    public UserIdentity getUserIdentity() {
+        return userIdentity;
+    }
+
+    public UserPrivilegeCollection getPrivilegeCollection() {
+        return privilegeCollection;
+    }
+
+    public short getPluginId() {
+        return pluginId;
+    }
+
+    public short getPluginVersion() {
+        return pluginVersion;
+    }
+
+    @Override
+    public void write(DataOutput out) throws IOException {
+        Text.writeString(out, GsonUtils.GSON.toJson(this));
+    }
+
+    public static UserPrivilegeCollectionInfoDeprecated read(DataInput in) throws IOException {
+        String json = Text.readString(in);
+        return GsonUtils.GSON.fromJson(json, UserPrivilegeCollectionInfoDeprecated.class);
+    }
+}

--- a/fe/fe-core/src/test/java/com/starrocks/privilege/AuthorizationMgrTest.java
+++ b/fe/fe-core/src/test/java/com/starrocks/privilege/AuthorizationMgrTest.java
@@ -30,6 +30,7 @@ import com.starrocks.meta.MetaContext;
 import com.starrocks.persist.OperationType;
 import com.starrocks.persist.RolePrivilegeCollectionInfo;
 import com.starrocks.persist.UserPrivilegeCollectionInfo;
+import com.starrocks.persist.UserPrivilegeCollectionInfoDeprecated;
 import com.starrocks.persist.gson.GsonUtils;
 import com.starrocks.persist.metablock.SRMetaBlockEOFException;
 import com.starrocks.persist.metablock.SRMetaBlockException;
@@ -1863,5 +1864,35 @@ public class AuthorizationMgrTest {
         ShowGrantsStmt showStreamLoadStmt = (ShowGrantsStmt) UtFrameUtils.parseStmtWithNewParser(sql, ctx);
         ShowExecutor executor = new ShowExecutor(ctx, showStreamLoadStmt);
         ShowResultSet resultSet = executor.execute();
+    }
+
+    @Test
+    public void testUserPriUpgrade() throws IOException, InterruptedException, PrivilegeException {
+        UtFrameUtils.PseudoJournalReplayer.resetFollowerJournalQueue();
+        UtFrameUtils.PseudoImage emptyImage = new UtFrameUtils.PseudoImage();
+
+        UserPrivilegeCollection collection = new UserPrivilegeCollection();
+        HashSet<Long> roleIds = new HashSet<>();
+        roleIds.add(1L);
+        collection.grantRoles(roleIds);
+        collection.setDefaultRoleIds(roleIds);
+        collection.grant(ObjectTypeDeprecate.SYSTEM, Lists.newArrayList(PrivilegeType.OPERATE),
+                Lists.newArrayList(new PEntryObject[] {null}), false);
+
+        UserPrivilegeCollectionInfoDeprecated deprecated = new UserPrivilegeCollectionInfoDeprecated(
+                new UserIdentity("u1", "%"),
+                collection,
+                (short) 1, (short) 1);
+
+        deprecated.write(emptyImage.getDataOutputStream());
+
+        UserPrivilegeCollectionInfo userPrivilegeCollectionInfo =
+                UserPrivilegeCollectionInfo.read(emptyImage.getDataInputStream());
+
+        Assert.assertEquals(new UserIdentity("u1", "%"), userPrivilegeCollectionInfo.getUserIdentity());
+        Assert.assertEquals(roleIds, userPrivilegeCollectionInfo.getPrivilegeCollection().getAllRoles());
+        Assert.assertEquals(roleIds, userPrivilegeCollectionInfo.getPrivilegeCollection().getDefaultRoleIds());
+        Assert.assertTrue(userPrivilegeCollectionInfo.getPrivilegeCollection()
+                .check(ObjectType.SYSTEM, PrivilegeType.OPERATE, null));
     }
 }

